### DISCUSSION TITLE
Is it possible to use cyrilic filenames with spaces for filesystem data collections?

### DISCUSSION THREAD

**Discussion by HiRolla** [joomlatools/joomlatools-pages#877] [Q&A] [answered]
Is it possible to use cyrilic filenames with spaces for filesystem data collections? I'm going to publish my growing markdown wiki via Pages, add joomla-hosted comments and so on. I would like to keep the original file naming like:

```
en/Introducrion article.md
ru/Вводная статья.md
```

As I understand now, MD files should be placed to the data folder. Then I should create some wiki.html.php file that will produce article lists and seo and user frienly routes.

Something like 

`function translit_path($value)`

Or I can fill out some slug field in the YAML of every article file.

Before or after that I should somehow convert non-path'ed [[links]] to make them work. 

I'm still at the very beginning. Thank you for such cool Pages and even wiki about it!  Your project is inspiring!

Please, tell me some advises on how should I start.  Yesterday I've got my first experience on php coding to make my prefered commenting system available on Pages, and now it's much more flexible via regexp url filtering. My first 27 lines of php code!

If this succeeds, it can be a great use-case, I think. I would like to participate!

**Comment by johanjanssens** [comment, marked as answer]
Glad to hear you are finding your way. and congrats on those first 27 lines of php code. This is how it all starts, one line of code at a time. Here is some feedback for you:

### 1. Loading external md files

The best approach to load the `.md`files would indeed be to use the Data API. I have replied to your earlier question in detail here: https://github.com/joomlatools/joomlatools-pages/discussions/873 Also added some other options how this could work.

### 2. Handling slugs

Two possible approaches here, you can get the name of the file (see below) or you could add the slug to the frontmatter of your markdown file, then just grab it from there using the Data API.

### 3. Cyrilic filenames with spaces

I haven't tested that but in theory this should just work. Pages is only loading the files, if your file system support cyrilic filenames with spaces then it should just be able to load them. Best to test that.

The Data API transforms each file into a [ComPagesDataObject](https://github.com/joomlatools/joomlatools-pages/blob/master/code/data/object.php) which extends from:

-  [ComPagesObjectConfig](https://github.com/joomlatools/joomlatools-pages/blob/master/code/object/config/config.php) 
- [KObjectConfig](https://github.com/joomlatools/joomlatools-framework/blob/master/code/libraries/joomlatools/library/object/config/config.php)

The data objects allows you to handle the data as an array, but also tap into the additional functions and features it offers, making the whole very powerful.  A markdown data file, implements an additional: `getFilename()` method that returns the actual name of the file: https://github.com/joomlatools/joomlatools-pages/blob/master/code/object/config/frontmatter.php#L113

You should be able to do

```php
<?= data('/path/to/file.md')->getFilename() ?>
```

Note this only works for `.md` at the moment, might make this a more generic API. 

### 4. Transliterate 

There are different ways you can handle the cyrilic file name in url's:

1. Leave it as is. URL's are UTF-8 and they should support this just fine.

2. Transliterate to another format, you have different options here:

- Using: building Pages `slug()`template method
- Using: https://www.php.net/manual/en/transliterator.transliterate.php
- Using: `\Joomla\CMS\Application\ApplicationHelper::stringURLSafe($value)`

Or as part of our framework you have two filters:

- https://github.com/joomlatools/joomlatools-framework/blob/master/code/libraries/joomlatools/component/koowa/filter/alias.php
- https://github.com/joomlatools/joomlatools-framework/blob/master/code/libraries/joomlatools/library/filter/slug.php

in Pages you would do: 

Convert to ASCII uses the PHP Transliterator if available

```php
<?= slug('Вводная статья') ?>
```
Convert to URL based on Joomla SEF settings use Joomla::stringURLSafe()

```php
<?= object('filter.alias')->sanitize('Вводная статья') ?>
```

Note there is no `alias()` template function in pages, but you could easily add one yourself. 

See also: https://stackoverflow.com/questions/7461406/cyrillic-transliteration-in-php

Happy coding.

**Comment by HiRolla** [comment]
Hello, Johan!
I've tested
`<?= sprintf('Text from the cyrilic markdown file: %s', data('произвольное имя файла.md')->content); ?>`
This is fantastic for me. So quick and simple!

Next,
`<?= slug('произвольное имя файла.md') ?>` returned `proizvolnoe-ima-fajlamd`
I know, that i should exclude the file extension, but here I expected that the point `.` should be also replaced by `-`.

Also i tried 
`<?= sprintf('NAME of the cyrilic markdown file: %s', data('произвольное имя файла.md')->getFilename()); ?>`
, but got `Call to undefined method ComPagesDataObject::getFilename()`. I'm trying to understand what the thing is.

**Reply by johanjanssens** [reply]
1. The slug filter doesn't replace dots with '-' it only does this for spaces, to solve this you could do.

`<?= slug(str_replace('.', ' ', 'произвольное имя файла.md')) ?> `

2. You are correct, I forgot about this. The object is merged into a big data object and the `getFilename()` method doesn't exist there. This won't work indeed, would need to make the data api smart to include the filename in each data object. Will have a look into that, might be able to do that.

**Comment by HiRolla** [comment]
I'm not sure that I understood how to use it, but when I try 
`<?= object('filter.alias')->sanitize('Вводная статья') ?>`
I get
```
[KObjectExceptionNotFound](https://example.com/go-php#error_page)	[500]
Cannot load object from identifier: lib:filter.alias
.../libraries/joomlatools/library/object/manager/manager.php:734
729                 throw new KObjectExceptionNotInstantiated(
730                     'Cannot instantiate object from identifier: ' . $class
731                 );
732             }
733         }
734         else throw new KObjectExceptionNotFound('Cannot load object from identifier: '. $identifier);
735 
736         return $result;
737     }
738 }
```
On the other hand I get what I need by this:
`JFilterOutput::stringURLSafe('произвольное имя файла.md')`
`proizvolnoe-imya-fajla-md`

✅ That fits my needs on transliteration, so thank you for the tip!

**Reply by johanjanssens** [reply]
Try: `<?= object('com:koowa.filter.alias')->sanitize('Вводная статья') ?>`

This filter is part of the Koowa component, not the base framework. This should work.
